### PR TITLE
feat(core): Initial TestRunner service with basic test execution (no-changelog)

### PR DESCRIPTION
**Pull request by burivuhster**
## Summary

This PR introduces a new test runner service that enables execution of test workflows.

Key features:
- Creates a test case set from existing executions according to the criteria set in test definition (aka specific annotation tag applied)
- Triggers the execution of workflow under test with all trigger nodes data mocked with the values from each specific test case
- Awaits for the test execution to finish before continuing with the next test case

Additionally, the new endpoint was added to trigger running tests for a specific test definition.

## Related Linear tickets, Github issues, and Community forum posts

- https://linear.app/n8n/issue/AI-427/[testrunner]-kick-off-workflow-under-test

## Review / Merge checklist

- [ ] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) 
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [ ] Tests included. 
- [ ] PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)

### CODE REVIEW

File: packages/cli/src/evaluation/test-runner/test-runner.service.ee.ts [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+		const executionData = parse(execution.executionData.data) as IExecutionResponse['data'];
+
+		const triggerNodes = workflow.nodes.filter((node) => /trigger$/i.test(node.type));
+
+		const pinData = {} as IPinData;
+
+		for (const triggerNode of triggerNodes) {
+			const triggerData = executionData.resultData.runData[triggerNode.name];
+			if (triggerData?.[0]?.data?.main?.[0]) {
+				pinData[triggerNode.name] = triggerData[0]?.data?.main?.[0];
```
Comment (tomi) [inline packages/cli/src/evaluation/test-runner/test-runner.service.ee.ts:56]: Do we have any (AI?) trigger nodes where the data might be in some other "port" than `main`? I have no clue, probably not. Would you @OlegIvaniv know?

File: packages/cli/src/evaluation/test-runner/test-runner.service.ee.ts [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+	 * Runs a test case with the given pin data.
+	 * Waits for the workflow under test to finish execution.
+	 * @param workflow
+	 * @param testCase
+	 * @param userId
+	 * @private
+	 */
+	private async runTestCase(
+		workflow: WorkflowEntity,
+		testCase: IPinData,
```
Comment (tomi) [inline packages/cli/src/evaluation/test-runner/test-runner.service.ee.ts:73]: The parameter name and type have bit of a mismatch. Would `testData` be accurate? Or `testInputData`? Or something like that

File: packages/cli/src/evaluation/test-runner/test-runner.service.ee.ts [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+			partialExecutionVersion: '-1',
+			userId,
+		};
+
+		// Trigger the workflow under test with mocked data
+		const executionId = await this.workflowRunner.run(data);
+		assert(executionId);
+
+		// Wait for the workflow to finish execution
+		const executePromise = Container.get(ActiveExecutions).getPostExecutePromise(
```
Comment (tomi) [inline packages/cli/src/evaluation/test-runner/test-runner.service.ee.ts:90]: Could we use constructor DI instead to get `ActiveExecutions`?

File: packages/cli/src/evaluation/test-runner/test-runner.service.ee.ts [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+			executionId,
+		) as Promise<IExecutionDb | undefined>;
```
Comment (tomi) [inline packages/cli/src/evaluation/test-runner/test-runner.service.ee.ts:92]: Why the type assertion here? `getPostExecutePromise` returns `IRun | undefined`. Is this safe to do?

File: packages/cli/src/evaluation/test-runner/test-runner.service.ee.ts [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+			executionMode: 'evaluation',
+			runData: {},
+			pinData: testCase,
+			workflowData: workflow,
+			partialExecutionVersion: '-1',
+			userId,
+		};
+
+		// Trigger the workflow under test with mocked data
+		const executionId = await this.workflowRunner.run(data);
```
Comment (tomi) [inline packages/cli/src/evaluation/test-runner/test-runner.service.ee.ts:86]: Do we need a timeout for the run? Or do we allow it to run as long as is needed? Does user have the option to cancel test runs that might be "stuck"?

File: packages/cli/src/evaluation/test-runner/test-runner.service.ee.ts [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+import { WorkflowRunner } from '@/workflow-runner';
+
+/**
+ * This service orchestrates the running of test cases.
+ * It uses the test definitions to find
+ * past executions, creates pin data from them,
+ * and runs the workflow-under-test with the pin data.
+ * TODO: Evaluation workflows
+ * TODO: Node pinning
+ * TODO: Collect metrics
```
Comment (tomi) [inline packages/cli/src/evaluation/test-runner/test-runner.service.ee.ts:24]: Do we have an entity representing the single run of a test case? Or how are the results shown to the user?

File: packages/cli/src/evaluation/test-runner/__tests__/test-runner.service.ee.test.ts [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+import type { ExecutionRepository } from '@/databases/repositories/execution.repository';
+import type { WorkflowRepository } from '@/databases/repositories/workflow.repository';
+import { NotFoundError } from '@/errors/response-errors/not-found.error';
+import type { TestDefinitionService } from '@/evaluation/test-definition.service.ee';
+import type { WorkflowRunner } from '@/workflow-runner';
+
+import { TestRunnerService } from '../test-runner.service.ee';
+
+const wfUnderTestJson = JSON.parse(
+	readFileSync(path.join(__dirname, './mock-data/workflow.under-test.json'), { encoding: 'utf-8' }),
```
Comment (tomi) [inline packages/cli/src/evaluation/test-runner/__tests__/test-runner.service.ee.test.ts:21]: Does this work? AFAIK we [mock FS access](https://github.com/n8n-io/n8n/blob/master/packages/cli/test/setup-mocks.ts#L8) in the tests at global level

**Review by tomi — COMMENTED** [review]
Left a bunch of comments and questions. Don't have a good overall picture of the feature so some of them might dumb

File: packages/cli/src/evaluation/test-runner/test-runner.service.ee.ts (hunk 6 above)
Comment (burivuhster) [reply packages/cli/src/evaluation/test-runner/test-runner.service.ee.ts:24]: It will be added in the subsequent PRs, yes. There will be TestRun entity with the aggregated collected metrics

File: packages/cli/src/evaluation/test-runner/test-runner.service.ee.ts [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+	 * Creates a new test run for the given test definition.
+	 * @param user
+	 * @param testId
+	 * @param accessibleWorkflowIds
+	 */
+	public async runTest(user: User, testId: string, accessibleWorkflowIds: string[]): Promise<any> {
+		const test = await this.testDefinitionsService.findOne(testId, accessibleWorkflowIds);
+
+		if (!test) {
+			throw new NotFoundError('Test definition not found');
```
[the remark this answers is not in the document]
Comment (burivuhster) [reply packages/cli/src/evaluation/test-runner/test-runner.service.ee.ts:97]: True... What do you think would be the right approach here?

[the remark this answers is not in the document]
Comment (tomi) [reply packages/cli/src/evaluation/test-runner/test-runner.service.ee.ts:97]: I would move the validation that the WF exists to hte controller

File: packages/cli/src/evaluation/test-runner/test-runner.service.ee.ts (hunk 5 above)
Comment (burivuhster) [reply packages/cli/src/evaluation/test-runner/test-runner.service.ee.ts:78]: We planned to implement an option for user to cancel stuck test runs, but will do this in a separate PR

File: packages/cli/src/evaluation/test-runner/__tests__/test-runner.service.ee.test.ts (hunk 7 above)
Comment (burivuhster) [reply packages/cli/src/evaluation/test-runner/__tests__/test-runner.service.ee.test.ts:18]: It seems we mock `node:fs`, but not the `fs`, this is why it works 🤔

File: packages/cli/src/evaluation/test-runner/test-runner.service.ee.ts (hunk 1 above)
Comment (burivuhster) [reply packages/cli/src/evaluation/test-runner/test-runner.service.ee.ts:52]: We decided to deal with proper node data pinning in a separate PR, as it has lots of edge cases like the one you mentioned. Here is just a "proof of concept", pinning the trigger node's main output.

**Review by tomi — APPROVED** [review]
LGTM now 🚀

**Review by OlegIvaniv — APPROVED** [review]

**Review by OlegIvaniv — APPROVED** [review]
Re-approving to trigger e2e pipeline

**Comment by janober** [comment]
Got released with [`n8n@1.70.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.70.0)
